packages/aws-cdk/lib/api/hotswap-deployments.ts

```diff
@@ -63,8 +63,62 @@ async function findAllHotswappableChanges(
   let foundNonHotswappableChange = false;
   const promises: Array<Array<Promise<ChangeHotswapResult>>> = [];
 
+  // as a first pass, we need to collapse replacement changes into one,
+  // as they are represented here as a pair of two changes: one addition and one removal
+  const allResourceChanges: { [logId: string]: cfn_diff.ResourceDifference } = stackChanges.resources.changes;
+  const allRemovalChanges = Object.entries(allResourceChanges).reduce((acc, [logId, resChange]) => {
+    if (resChange.isRemoval) { acc[logId] = resChange; }
+    return acc;
+  }, {} as { [logId: string]: cfn_diff.ResourceDifference });
+  const allNonRemovalChanges = Object.entries(allResourceChanges).reduce((acc, [logId, resChange]) => {
+    if (!resChange.isRemoval) { acc[logId] = resChange; }
+    return acc;
+  }, {} as { [logId: string]: cfn_diff.ResourceDifference });
+  const allResourceChangesWithReplacements: { [logId: string]: cfn_diff.ResourceDifference } = {};
+  for (const [logId, nonRemovalChange] of Object.entries(allNonRemovalChanges)) {
+    if (nonRemovalChange.isAddition) {
+      const addChange = nonRemovalChange;
+      // search for an identical removal change
+      const identicalRemovalChange = Object.entries(allRemovalChanges).find(([_, remChange]) => {
+        return remChange.oldResourceType === addChange.newResourceType &&
+          // this isn't great, but I don't want to bring in something like underscore just for this comparison
+          JSON.stringify(remChange.oldProperties) === JSON.stringify(addChange.newProperties);
+      });
+      // if we found one, this means this is a replacement change
+      if (identicalRemovalChange) {
+        allResourceChangesWithReplacements[logId] = new cfn_diff.ResourceDifference(
+          // we have to fill in the old value, because otherwise this will be classified as a non-hotswappable change
+          identicalRemovalChange[1].oldValue,
+          addChange.newValue,
```

Comment:
Aren't `oldValue` and `newValue` going to be identical by definition here?

Reply:
Yes. I still feel like using the different values is clearer though (same in the lines below). What do you think?

Reply:
Maybe. Once you factor it out to a function I care less.